src/execution_engine.rs

```diff
@@ -253,3 +314,61 @@ impl Drop for ExecutionEngine {
         }
     }
 }
+
+/// A wrapper around a function pointer which ensures the symbol being pointed
+/// to doesn't accidentally outlive its execution engine.
+#[derive(Clone)]
+pub struct Symbol<F> {
+    pub(crate) execution_engine: Rc<LLVMExecutionEngineRef>,
+    inner: F,
+}
+
+impl<F: UnsafeFunctionPointer> Deref for Symbol<F> {
+    type Target = F;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<F> Debug for Symbol<F> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_tuple("Symbol")
+            .field(&"<unnamed>")
```

Comment:
Maybe we should store the function name from `get_function` for debugging purposes?

Reply:
Hmm... I'm not sure whether that'll be nice to do. We're keeping track of the execution engine by using an `Rc`, and mixing both a reference (with lifetimes) and a reference-counted pointer feels kinda weird.

It also means your function name must outlive the symbol, which isn't really practical or ergonomic. Imagine trying to store a loaded symbol in a struct when the name isn't known until runtime (e.g. the user provided it), it'd be way too easy to get stuck in the rabbit hole that is self-referential types.

Reply:
The &str is tied to the input to the function call, not the execution engine though